modules/ocf_puppet/files/puppet-report

```diff
@@ -0,0 +1,181 @@
+#!/usr/bin/env python3
+"""A custom batch report processor for Puppet.
+
+Features of the processor
+* Doesn't email reports for nodes not on the production environment
+* Can send reports for multiple hosts in a single digest-like report
+  with the '--digest' option
+* Doesn't email reports which only contain the "switching agent
+  environment" message
+* Sends puppet reports for everything else like tagmail does
+* Clears processed logs from the master
+
+Failed builds in production will be reported by email. Failed builds on
+other environments will not, but will still be reported via IRC.
+
+The default options will capture more-or-less the default behavior of
+Puppet's tagmail. Since desktops all have essentially the same
+configuration, an improved setup is to process desktops and servers
+separately by specifying the 'desktop' and 'server' types (respectively)
+and adding the '--digest' option for desktops.
+
+The '--report-dir' option can be used to specify where to find Puppet
+reports on the master, should it differ from the default.
+
+Processing all this YAML is slow. It would be an improvement to
+parallelize this script.
+"""
+import os
+import sys
+from argparse import ArgumentParser
+from collections import namedtuple
+from datetime import datetime
+from datetime import timedelta
+from glob import glob
+
+import yaml
+from ocflib.infra.hosts import hostname_from_domain
+from ocflib.infra.hosts import type_of_host
+from ocflib.misc.mail import send_mail
+
+
+DEFAULT_TYPES = ['server', 'desktop']
+DEFAULT_REPORT_DIR = '/var/lib/puppet/reports'
+MAIL_TO_FROM = 'Puppet <puppet@ocf.berkeley.edu>'
+
+
+# Boilerplate to parse Puppet objects tagged in YAML
+
+PuppetReport = namedtuple('PuppetReport', ['logs', 'host', 'time',
+                                           'environment', 'status'])
+PuppetLog = namedtuple('PuppetLog', ['level', 'tags', 'message', 'source',
+                                     'time'])
+
+
+def NamedtupleConstructor(cls):
+    def RealConstructor(loader, node):
+        data = loader.construct_mapping(node)
+        return cls(*[data[field] for field in cls._fields])
+    return RealConstructor
+
+yaml.add_constructor(
+    u'!ruby/object:Puppet::Transaction::Report',
+    NamedtupleConstructor(PuppetReport),
+    Loader=yaml.SafeLoader
+)
+yaml.add_constructor(
+    u'!ruby/object:Puppet::Util::Log',
+    NamedtupleConstructor(PuppetLog),
+    Loader=yaml.SafeLoader
+)
+
+# Treat all other tagged objects like plain YAML
+# Do this by removing the default constructor, which throws an error
+del yaml.SafeLoader.yaml_constructors[None]
```

Comment:
Sure, there are multiple ways. This was just the absolute most
straightforward.
On May 27, 2016 6:08 PM, "Chris Kuehl" <notifications@github.com> wrote:

> In modules/ocf_puppet/files/puppet-report
> <https://github.com/ocf/puppet/pull/52#discussion_r64981365>:
>
> > +    return RealConstructor
> > +
> > +yaml.add_constructor(
> > +    u'!ruby/object:Puppet::Transaction::Report',
> > +    NamedtupleConstructor(PuppetReport),
> > +    Loader=yaml.SafeLoader
> > +)
> > +yaml.add_constructor(
> > +    u'!ruby/object:Puppet::Util::Log',
> > +    NamedtupleConstructor(PuppetLog),
> > +    Loader=yaml.SafeLoader
> > +)
> > +
> > +# Treat all other tagged objects like plain YAML
> > +# Do this by removing the default constructor, which throws an error
> > +del yaml.SafeLoader.yaml_constructors[None]
>
> this (and the stuff above) is kinda sad, this has global impact within the
> python process (i.e. yaml in another module is impacted). is there any
> way to avoid that?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ocf/puppet/pull/52/files/55cd8a4b2941f57596bec5c891261272c5f38df2#r64981365>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AGaTP9aTTcirVnx9e-YNQs5blGKdq9xQks5qF5WogaJpZM4Io2eC>
> .
>